gh-342 Sign in to SDE as a researcher using OpenID Connect

pjmonks opened the pull request:
Resolves #342 

**Note:** this PR must be reviewed alongside:

- MauroDataMapper/sde-core#114
- MauroDataMapper/sde-resources#75

These PRs should be accepted and merged first.

- Fix OpenID Connect authorization for Mauro sign in
  - Correctly redirects to dashboard page and updates header state with signed in user
  - Remove fake SDE sign in call, not required
- Create SDE module for components
- Adjust existing SDE main page to use new components
- Create sign in component for SDE researcher. Lists all available OIDC providers from SDE to trigger redirect and authentication
- Add redirect page for SDE authentication to complete login process
  - Micronaut backend will redirect to this page to complete SDE login
  - Successful login from OIDC will fetch user details of logged in user and put them in local storage
  - Redirect to correct page and display user details
  - Update SdeRestHandler to use `withCredentials` option to always send authorization cookies back to SDE backend
- Sign in is automatic when Mauro SSO is used, with the proviso that:
  - Mauro and SDE OpenID Connect providers are configured exactly the same.
  - Providers share the same labels (to synchronise them)
  - Then SDE sign in occurs automatically as soon as Mauro sign in succeeds
 - Sign out is automatic when Mauro SSO is used:
   - A hidden HTML form had to be put on the `app.component`, which is where the original Mauro sign out event is handled
   - Once Mauro is signed out, the hidden HTML form is automatically submitted to trigger a HTTP POST request to the SDE

Notes about testing:

1. OCC logins via Azure AD mean we have two accounts available for SSO - OCC and SystemC. Because of the way I had to implement the SSO login loop - login to Mauro session first, then automatically login to SDE session - you will see the Azure account selection page twice to select an account (pick OCC both times). This duplication is not desirable but needs to be addressed at a later date.

pjmonks committed:
gh-342 Update to support OpenID Connect authentication for SDE researcher

- Fix OpenID Connect authorization for Mauro sign in
  - Correctly redirects to dashboard page and updates header state with signed in user
  - Remove fake SDE sign in call, not required
- Create SDE module for components
- Adjust existing SDE main page to use new components
- Create sign in component for SDE researcher. Lists all available OIDC providers from SDE to trigger redirect and authentication
- Add redirect page for SDE authentication to complete login process
  - Micronaut backend will redirect to this page to complete SDE login
  - Successful login from OIDC will fetch user details of logged in user and put them in local storage
  - Redirect to correct page and display user details
  - Update SdeRestHandler to use `withCredentials` option to always send authorization cookies back to SDE backend
- Add method to sign out user of SDE backend
- Display minimal information on SDE main page once signed in to prove the process has worked

pjmonks committed:
gh-342 Fix tests and linting

pjmonks committed:
gh-342 Adjust single sign on so that Mauro and SDE both sign in/out at the same time

- Sign in requires chaining OpenID Connect providers together so one happens after another automatically
- Sign out happens through a hidden HTML form to sign out from SDE. The sign out event, when completed for Mauro, then automatically submits the form to trigger a POST request to SDE

pjmonks committed:
gh-342 Fix tests to pass after changes to SDE single sign-on

pjmonks committed:
Merge branch 'develop' into feature/gh-342-sde-login-oidc

greg-at-occ commented on src/app/app.component.ts:
```diff
@@ -283,19 +297,13 @@ export class AppComponent implements OnInit, OnDestroy {
+        if (this.sdeSignOutForm) {
+          // If SDE sign out form is available, trigger a form POST request
+          // This will force the app to sign out of the SDE too automatically
+          this.sdeSignOutForm.nativeElement.submit();
```
Is it not possible to do this sign out with a direct XHR call? I'm curious why we have this hidden form approach.

pjmonks replied:
Unfortunately not. I've updated this code with a more detailed comment explaining this bizarre workaround.

greg-at-occ commented on src/app/app.component.ts:
```diff
@@ -184,7 +198,8 @@ export class AppComponent implements OnInit, OnDestroy {
     private error: ErrorService,
-    private themes: ThemeService
+    private themes: ThemeService,
+    private sdeAuthentication: AuthenticationEndpoints
```
I think we will want to move this into sde-resources

pjmonks replied:
This has been done now.

pjmonks committed:
gh-342 Move SDE authentication endpoints to sde-resources

pjmonks committed:
gh-342 Fix tests

greg-at-occ commented on src/app/secure-data-environment/components/organisation-list/authentication.resources.ts:
```diff
@@ -0,0 +1,53 @@
+  imageUrl?: string;
+}
+
+export const sdeOpenIdConnectProviders: SdeOpenIdConnectProvider[] = [
```
This is in sde-resources now, should it be removed here?

pjmonks replied:
Yes, I've removed this obsolete copy.

greg-at-occ commented on src/app/secure-data-environment/resources/authentication.resources.ts:
```diff
@@ -0,0 +1,47 @@
+/*
```
There are two copies of authentication.resources.ts in different places here, plus another in sde-resources. Can we dispose of some of them?

pjmonks replied:
Yes, I've removed this obsolete copy.

greg-at-occ commented on src/app/app.component.ts:
```diff
@@ -42,14 +48,20 @@ import { FooterLink } from './shared/footer/footer.component';
+   * Hidden HTML form element to manually trigger a sign out of the SDE.
+   * This is automatically handled in the broadcast 'sign-out-user' event.
+   */
+  @ViewChild('sdeSignOutForm') sdeSignOutForm?: ElementRef<HTMLFormElement>;
```
I think I commented on this before but I can't find my comment right now: do we need a hidden form for this? Is there not an XHR approach?

pjmonks replied:
Unfortunately not. I've updated this code with a more detailed comment explaining this bizarre workaround.

greg-at-occ commented on src/app/pages/sign-in/sign-in.component.ts:
```diff
@@ -80,10 +82,29 @@ export class SignInComponent implements OnInit {
+      .pipe(
+        catchError(() => []),
+        map((sdeProviders) =>
+          // Big assumption - that Mauro and SDE have OpenID Connect providers are configured
```
Can we use a more reliable detail than the label like the URI for the IDP?

pjmonks replied:
Unfortunately I don't think we can at the moment, it's a problem with storing the same Oauth provider details in two different systems. I've elaborated more in the comment instead.

greg-at-occ replied:
In other words we would need to make changes to the MDM core OIDC plugin to support a more robust approach?

greg-at-occ commented on src/app/secure-data-environment/sde-sign-in/sde-sign-in.component.html:
```diff
@@ -0,0 +1,62 @@
+<!--
```
Possibly combine with equivalent component in sde-admin-ui?

pjmonks replied:
I've now changed this to use a shared component from `sde-resources`

greg-at-occ commented on src/app/security/user-details.service.ts:
```diff
@@ -105,4 +106,37 @@ export class UserDetailsService {
+      id: localStorage.getItem('sdeUserId') ?? '',
+      email: localStorage.getItem('sdeEmail') ?? '',
+      preferredName: localStorage.getItem('sdePreferredName') ?? '',
+      isDeleted: false,
```
Is filling in all these fields with default values a sign of a problem elsewhere?

pjmonks replied:
This function wasn't strictly necessary so I've removed it instead.

greg-at-occ reviewed (changes requested):
A bit of tidy up, and could you confirm that you decided against combining a couple of components into sde-resources?

pjmonks committed:
gh-342 Use sde-resources shared component for using SDE oauth provider button list

pjmonks committed:
gh-342 Move authentication finalization to sde-resources

- Keep page level component to handle sign in/out events, but host the shared component
- Update tests

pjmonks committed:
gh-342 Clean up code

- Remove obsolete code files
- Add comments to app.component to explain why sign out of SDE goes through a browser workaround

pjmonks committed:
gh-342 Further updates from PR feedback

- Add better comment to explain why labels of Oauth providers are needed to sync same provider between Mauro and SDE
- Remove UserDetails.getSdeResearchUser() - was not needed

greg-at-occ reviewed (approved):
This is great and works so smoothly
